pinot-tools/src/main/java/com/linkedin/pinot/tools/ClusterStateVerifier.java

```diff
@@ -74,18 +77,42 @@ public Boolean call() {
    * @param tableNames list of table names which are about to scan
    * @return
    */
-  private boolean isClusterStable(final List<String> tableNames) {
-    LOGGER.info("Start scanning the stability of all the tables... Num of tables: {}", tableNames.size());
-    int numUnstablePartitions = 0;
-    for (String tableName : tableNames) {
-      numUnstablePartitions = super.isStable(tableName);
-      if (numUnstablePartitions != 0) {
-        LOGGER.error("Attention: Table {} is not stable. numUnstablePartitions: {} ", tableName, numUnstablePartitions);
-        break;
+  private boolean waitForClusterStable(final List<String> tableNames, final long timeoutSec)
+      throws InterruptedException{
+    final long startTimeMillis = System.currentTimeMillis();
+    final long maxEndTimeMillis = startTimeMillis + TimeUnit.MILLISECONDS.convert(timeoutSec, TimeUnit.SECONDS);
+
+    boolean stable = false;
+    int iteration = 0;
+    long sleepTimeMillis = (maxEndTimeMillis - startTimeMillis)/10;
+    if (sleepTimeMillis < MIN_SLEEP_BETWEEN_CHECKS_MILLIS) {
+      sleepTimeMillis = MIN_SLEEP_BETWEEN_CHECKS_MILLIS;
+    } else if (sleepTimeMillis > MAX_SLEEP_BETWEEN_CHECKS_MILLIS) {
+      sleepTimeMillis = MAX_SLEEP_BETWEEN_CHECKS_MILLIS;
+    }
+
+    while (!stable) {
+      iteration++;
+      LOGGER.info("Start scanning the stability of {} tables, iteration {}",
+          tableNames.size(), iteration);
+      int numUnstablePartitions = 0;
+      for (String tableName : tableNames) {
+        numUnstablePartitions = super.isStable(tableName);
+        if (numUnstablePartitions != 0) {
+          LOGGER.error("Table {} is not stable. numUnstablePartitions: {} ", tableName, numUnstablePartitions);
+          break;
```

Comment:
I've limited context for this tool and concept of stability. This check is not atomic (a checked table may turn unstable by the time this processes last table). Given that, in case of some unstable tables, does it make sense to check all tables and truncate already validated tables before sleeping ?  That will depend on the total time taken to check for stability.

Reply:
Correct, this may happen.The intention here is that all tables reach stability, so we are OK. Worst case, we time out, and there is a message that says which tables are not stable, so it is ok.